File: Algorithms/src/1340. Jump Game V/Code/solution.js
/**
 * @param {number[]} arr
 * @param {number} d
 * @return {number}
 */
var maxJumps = function (arr, d) {
  const n = arr.length;

  // dp[i] stores maximum jumps starting from i
  const dp = new Array(n).fill(-1);

  // DFS function
  const dfs = (i) => {
    // Return stored answer
    if (dp[i] !== -1) return dp[i];

    // Current index counts as 1
    let ans = 1;

    // Move right
    for (let j = i + 1; j <= Math.min(n - 1, i + d); j++) {
      // Stop if blocked
      if (arr[j] >= arr[i]) break;

      // Update answer
      ans = Math.max(ans, 1 + dfs(j));
    }

    // Move left
    for (let j = i - 1; j >= Math.max(0, i - d); j--) {
      // Stop if blocked
      if (arr[j] >= arr[i]) break;

      // Update answer
      ans = Math.max(ans, 1 + dfs(j));
    }

    // Store answer
    dp[i] = ans;

    return ans;
  };

  let answer = 1;

  // Start from every index
  for (let i = 0; i < n; i++) {
    answer = Math.max(answer, dfs(i));
  }

  return answer;
};

File: Algorithms/src/1340. Jump Game V/Code/solution.go
func maxJumps(arr []int, d int) int {

	n := len(arr)

	// dp[i] stores maximum jumps starting from i
	dp := make([]int, n)

	// Initialize with -1 meaning unvisited
	for i := 0; i < n; i++ {
		dp[i] = -1
	}

	// DFS function
	var dfs func(int) int

	dfs = func(i int) int {

		// Return stored answer
		if dp[i] != -1 {
			return dp[i]
		}

		// Current index itself counts as 1
		ans := 1

		// Move right
		for j := i + 1; j <= min(n-1, i+d); j++ {

			// Stop if blocked
			if arr[j] >= arr[i] {
				break
			}

			// Update best answer
			ans = max(ans, 1+dfs(j))
		}

		// Move left
		for j := i - 1; j >= max(0, i-d); j-- {

			// Stop if blocked
			if arr[j] >= arr[i] {
				break
			}

			// Update best answer
			ans = max(ans, 1+dfs(j))
		}

		// Store result
		dp[i] = ans

		return ans
	}

	answer := 1

	// Try every starting index
	for i := 0; i < n; i++ {
		answer = max(answer, dfs(i))
	}

	return answer
}

// Helper function for maximum
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Helper function for minimum
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}